File: merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> res;
        if(intervals.size()==1)
        {
            res.push_back(intervals[0]);
            return res;
        }
        sort(intervals.begin(),intervals.end());
        if(intervals[1][0]<=intervals[0][1])
            res.push_back(vector<int>{intervals[0][0],max(intervals[1][1],intervals[0][1])});
        else
        {
            res.push_back(intervals[0]);
            res.push_back(intervals[1]);
        }
        for(int i=2;i<intervals.size();i++)
        {
            int n=res.size();
            if(intervals[i][0]<=res[n-1][1])
                res[n-1][1]=max(intervals[i][1],res[n-1][1]);
            else
                res.push_back(intervals[i]);
        }
        return res;
    }
};

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        mp = defaultdict(int)
        for interval in intervals:
            mp[interval[0]] += 1
            mp[interval[1]] -= 1
        mp_key = sorted(mp.keys())
        prefix_sum = 0
        ans = []
        tmp_interval = []
        for k in mp_key:
            if prefix_sum == 0:
                tmp_interval.append(k)
            prefix_sum += mp[k]
            mp[k] = prefix_sum
            if mp[k] == 0:
                tmp_interval.append(k)
                ans.append(tmp_interval)
                tmp_interval = []
        return ans